src/client/ds3/bucket.c

```diff
@@ -0,0 +1,419 @@
+{
+	int rc = 0;
+
+	rc = dfs_disconnect(ds3b->dfs);
+	D_FREE(ds3b);
+	return -rc;
+}
+
+int
+ds3_bucket_get_info(struct ds3_bucket_info *info, ds3_bucket_t *ds3b, daos_event_t *ev)
+{
+	int               rc       = 0;
+	char const *const names[]  = {RGW_BUCKET_INFO};
+	void *const       values[] = {info->encoded};
+	size_t            sizes[]  = {info->encoded_length};
+	daos_handle_t     coh;
+
+	if (ds3b == NULL || info == NULL)
+		return -EINVAL;
+
+	rc = dfs_cont_get(ds3b->dfs, &coh);
+	if (rc != 0) {
+		return -rc;
+	}
+
+	rc = daos_cont_get_attr(coh, 1, names, values, sizes, ev);
+	rc = daos_der2errno(rc);
+	dfs_cont_put(ds3b->dfs, coh);
+	return -rc;
+}
+
+int
+ds3_bucket_set_info(struct ds3_bucket_info *info, ds3_bucket_t *ds3b, daos_event_t *ev)
+{
+	int               rc       = 0;
+	char const *const names[]  = {RGW_BUCKET_INFO};
+	void const *const values[] = {info->encoded};
+	size_t const      sizes[]  = {info->encoded_length};
+	daos_handle_t     coh;
+
+	if (ds3b == NULL || info == NULL)
+		return -EINVAL;
+
+	rc = dfs_cont_get(ds3b->dfs, &coh);
+	if (rc != 0) {
+		return -rc;
+	}
+
+	rc = daos_cont_set_attr(coh, 1, names, values, sizes, ev);
+	rc = daos_der2errno(rc);
+	dfs_cont_put(ds3b->dfs, coh);
+	return -rc;
+}
+
+int
+ds3_bucket_list_obj(uint32_t *nobj, struct ds3_object_info *objs, uint32_t *ncp,
+		    struct ds3_common_prefix_info *cps, const char *prefix, const char *delim,
+		    char *marker, bool list_versions, bool *is_truncated, ds3_bucket_t *ds3b)
+{
+	int            rc = 0;
+	char          *file_start;
+	const char    *path        = "";
+	const char    *prefix_rest = prefix;
+	dfs_obj_t     *dir_obj;
+	char          *lookup_path;
+	struct dirent *dirents;
+	daos_anchor_t  anchor;
+	uint32_t       cpi  = 0;
+	uint32_t       obji = 0;
+	uint32_t       i;
+	const char    *name;
+	dfs_obj_t     *entry_obj;
+	mode_t         mode;
+	char          *cpp;
+
+	if (ds3b == NULL || nobj == NULL)
+		return -EINVAL;
+
+	// End
+	if (*nobj == 0)
+		return 0;
+
+	// TODO: support the case when delim is not /
+	if (strcmp(delim, "/") != 0) {
+		return -EINVAL;
+	}
+
+	file_start = strrchr(prefix, delim[0]);
+	if (file_start != NULL) {
+		*file_start = '\0';
+		path        = prefix;
+		prefix_rest = file_start + 1;
+	}
+
+	D_ALLOC_ARRAY(lookup_path, DS3_MAX_KEY);
+	if (lookup_path == NULL)
+		return -ENOMEM;
+
+	strcpy(lookup_path, "/");
+	strcat(lookup_path, path);
+	rc = dfs_lookup(ds3b->dfs, lookup_path, O_RDWR, &dir_obj, NULL, NULL);
+	if (rc != 0) {
+		goto err_path;
+	}
+
+	D_ALLOC_ARRAY(dirents, *nobj);
+	if (dirents == NULL) {
+		rc = ENOMEM;
+		goto err_dir_obj;
+	}
+
+	// TODO handle bigger directories
+	// TODO handle ordering
+	// TODO handle marker
+	daos_anchor_init(&anchor, 0);
+
+	rc = dfs_readdir(ds3b->dfs, dir_obj, &anchor, nobj, dirents);
+	if (rc != 0) {
+		goto err_dirents;
+	}
+
+	if (is_truncated != NULL) {
+		*is_truncated = !daos_anchor_is_eof(&anchor);
+	}
+
+	for (i = 0; i < *nobj; i++) {
```

Comment:
it would help to add some comments on what this is trying to do so i could tell if this is the best way to do accomplish this. im afraid im not getting it